### ISSUE TITLE
Is there a way to enter a predefined secret ?

### ISSUE THREAD

**Issue by xavgra2** [j81blog/OTP4ADC#2]
Hi, 
I have a special use case where I need to add a hardware token for a user, and the secret is hardcoded in the device.
So it will be very nice if there will be an option to enter manually the "secret" instead of "auto-generate" it when adding one.

Do you see what I mean ? of course, I'll be pleased to help if possible.

**Comment by j81blog** [comment]
Not sure if this will work. I don't know if the same method is being used in generating the codes. But you will need the seed string. You can try and test it yourself? (although without encryption, with the "cleartext" option. You just save your own data to the active directory attribute of a user. You can enter the following string:
```
#@DeviceName=SECRET&,
```
'DeviceName' can be any name.
Replace 'SECRET' with your seed string. 
Make sure the string begins with '#@' and ends with '&,'
Then you can test it.

Please let me know if it works, then I can make the necessary changes in the tool to support it.
I don't have any hardware tokens, so I can't test it.

**Comment by xavgra2** [comment]
Because I’m not comfortable in editing the field in the AD directly, I propose to give you more details ;
One of the provider I found, but it’s not the only one, offer a toolbox to play with totp.
You can fin dit here : 
https://www.token2.com/site/page/totp-toolset

When you buy a hardware device, the supplier will give you a file with the corresponding TOTP seed encoded in base32.
You can easily use the link above to create a « virtual-hardware device ».

Does it help ?
Or do you still need me to do the test you requested ?

**Comment by j81blog** [comment]
I've added an option to allow you to add your own secret. To use this option, disable the checkbox for "Change Secret". You can find the updated version in [=> Dev <=](https://github.com/j81blog/OTP4ADC/tree/dev)
Please let me know the results. My initial tests looks promising.

**Comment by xavgra2** [comment]
We just did the test and it works perfectly well. Thank you for your support and your tool .

**Comment by token2** [comment]
> I've added an option to allow you to add your own secret. To use this option, disable the checkbox for "Change Secret". You can find the updated version in [=> Dev <=](https://github.com/j81blog/OTP4ADC/tree/dev) Please let me know the results. My initial tests looks promising.

Hi, we can also suggest adding an option to bulk update the base32 secrets. Ideally from a CSV file , something similar to what Azure MFA is using:
`upn,serial number,secret key,time interval,manufacturer,model
Helga@contoso.com,1234567,2234567abcdef2234567abcdef,60,Contoso,HardwareKey`

**Comment by j81blog** [comment]
"Bulk" options are already available via command-line, I will also add an option to allow adding secrets via command-line.

**Comment by j81blog** [comment]
With the 1.0.1 dev version you can now set secrets via commandline. You could do something like:
```PowerShell
$exportPath = "C:\export"
$gatewayURI = "portal.domain.com"
$attribute = "userParameters"
$csvData = @"
"Username","DeviceName","Secret"
"john.doe@domain.com","Mobile","NMBIP7SKCPOFSVWUYD36VMJISM"
"jane.doe@domain.com","Mobile","NMBIP7SKCPOFSVWUYD36VMJISM"
"@ | ConvertFrom-CSV
New-Item -Path $exportPath -ItemType Directory -Force | Out-Null
$results = @()
ForEach ($item in $csvData) {
    $results += C:\Scripts\OTP4ADC.ps1  -attribute $attribute -GatewayURI $gatewayURI -username $item.Username -DeviceName $item.DeviceName -ExportPath $exportPath    
}
$results
```

**Comment by token2** [comment]
> ```powershell
> -Path $exportPath
> ```
Thanks for this update. Can you please clarify if we can skip the -Path $exportPath parameter completely as setting secrets is not expected to produce any export file.   If I do not use this key it asks for it in when executing.  Setting -ExportPath disabled, null or false creates folders with such names

**Comment by j81blog** [comment]
I made a new dev version v1.0.3, I haven't had much time to test it. it won't need an export path if you specify a secret. I also added an option to directly import a CSV. 

if you use a CSV use the FileImport parameter 

```Powershell
.\OTP4ADC.ps1 -CsvPath C:\temp\users.csv -ReplaceTokens -GatewayURI portal.domain.com -Attribute userparameters
```

users.csv (example)
```csv
"Username","DeviceName","Secret"
"john.doe@domain.com","Mobile","ABCDEF"
```

Please let me know how it goes.

**Comment by token2** [comment]
> I made a new dev version v1.0.3, I haven't had much time to test it. it won't need an export path if you specify a secret. I also added an option to directly import a CSV.
> 
> if you use a CSV use the FileImport parameter
> 
> ```powershell
> .\OTP4ADC.ps1 -CsvPath C:\temp\users.csv -ReplaceTokens -GatewayURI portal.domain.com -Attribute userparameters
> ```
> 
> users.csv (example)
> 
> ```
> "Username","DeviceName","Secret"
> "john.doe@domain.com","Mobile","ABCDEF"
> ```
> 
> Please let me know how it goes.

We tested with the previous version with a workaround of removing the ExportPath :
https://www.token2.swiss/site/page/using-classic-tokens-for-native-otp-authentication-with-netscaler

Works well, a couple of customers already tested and confirmed. Will test the latest version and update 

Thanks and happy holidays!

**Comment by j81blog** [comment]
I published [v1.0.3](https://github.com/j81blog/OTP4ADC/releases/tag/v1.0.3) and changed the readme accordingly.

**Comment by xavgra2** [comment]
Thank you very much for these very usefull change !
